plutus-core/src/Language/PlutusCore/Constant/Typed.hs

```diff
@@ -168,23 +184,61 @@ with the following signature:
+in doing so as GHC can infer all the kinds itself
+3. ... apart from cases where they're inherently ambiguous, like in the case above. If we don't
+specify the kind of the @a_1@ type variable, then there's no way GHC could infer it as the variable
+is passed as an argument to another variable with an unspecified kind (@f_0@)
+4. finally, an opaque term can only be of type of kind @*@. You can't construct a term whose type
+is of some other kind. That's why we don't need to annotate the @f_0@ type variable: the domain is
+inferred from the kind of the argument (where it's explicitly specified via @TyNameRep *@) and the
+codomain is inferred from the fact that the whole type is passed to 'Opaque' and so it has to be
+of kind @*@
+
+It would be nice if we didn't need to define that @*Rep@ stuff at the type level just to demote it
+to the term level via a type class, but this hasn't been investigated yet. A plausible way would be
+to ditch 'KnownTypeAst' (but keep 'KnownType') and provide PLC types manually. But that doesn't seem
+to give rise to a terribly nice API. And we'd lose all the static guarantees, which is not a big
+deal, but losing the automatic inference of type schemes would suck, given that it's already quite
+handy and is going to be improved.
+
+Representing contructors as poly-kinded data families and handling those with open type families
+and/or type classes is a way of solving the expression problem for indexed data types at the type
+level, if you are into these things.
+-}
 
-It would be much better if we didn't need to define that @*Rep@ stuff at the type level just to
-demote it to the term level via a type class, but this hasn't been investigated yet. A plausible
-way would be to ditch 'KnownTypeAst' (but keep 'KnownType') and provide PLC types manually.
-But that doesn't seem to give rise to a terribly nice API.
+-- | Representation of a type variable: its name and unique and an implicit kind.
+data TyNameRep (kind :: GHC.Type) = TyNameRep Symbol Nat
 
-The implementation is rather straightforward, but there is one catch, namely that it takes some care
-to adapt the CEK machine to handle polymorphic built-in functions,
-see https://github.com/input-output-hk/plutus/issues/1882
--}
+-- | Representation of an intrinsically kinded type variable: a name.
+data family TyVarRep (var :: TyNameRep kind) :: kind
 
--- | Representation of a type variable: its name and unique.
-data TyVarRep (text :: Symbol) (unique :: Nat)
+-- | Representation of an intrinsically kinded type application: a function and an argument.
+data family TyAppRep (fun :: dom -> cod) (arg :: dom) :: cod
 
--- | Representation of a type application: a function and an argument.
-data TyAppRep (fun :: GHC.Type) (arg :: GHC.Type)
+-- | Representation of of an intrinsically kinded universal quantifier: a bound name and a body.
+data family TyForallRep (var :: TyNameRep kind) (a :: GHC.Type) :: GHC.Type
```

Comment:
BTW, what do you think about `StandaloneKindSignatures`? I think I like the fact that it's a very regular way to give signatures to things like this. So it would be more like:
```
type TyForallRep :: TyNameRep kind -> GHC.Type -> GHC.Type
data family TyForallRep var a 
```

Reply:
Don't have any strong opinion. I guess I just like keeping that definition a one-liner, but if you prefer a separate kind signature, I can add it.

Reply:
I haven't really used them much, I just read this recently: https://ryanglscott.github.io/2020/01/05/five-benefits-to-using-standalonekindsignatures/

Reply:
I guess, I'll randomly go with the current no-SAKS version, because I don't care and so choose the path of least effort (not changing anything).